### DISCUSSION TITLE
Coverage

### DISCUSSION THREAD

**Discussion by jbcaillau** [control-toolbox/control-toolbox.github.io#25] [Q&A]
From @j-l-s :  
> comment vous visualisez les tests coverage ? (en python, il y a un package qui tranforme les .cov en html et on peut naviguer dans les tableaux de tests/stats assez agréablement dans un browser web)

@ocots can you please answer?

**Comment by ocots** [comment]
we click on `codecov` at the root of the repo.

![Capture d’écran 2023-03-24 à 10 58 33](https://user-images.githubusercontent.com/66357348/227490362-b555308f-073c-4f94-bf7a-e4860b16afdd.png)

**Comment by j-l-s** [comment]
Mon soucis est d'écrire des tests qui couvrent le max de code. Les fichier .cov générés par `Pkg.test(coverage=true)` ne sont pas franchement sympathiques à utiliser. Du coup, je suis dans ma branche, sur ma machine.... 
(ceci-dit ça aide)

**Reply by ocots** [reply]
J'ai jamais testé mais avec VSCode : https://discourse.julialang.org/t/code-coverage-with-coverage-jl-in-vscode/65438/3#julia-line-coverage-information-in-vs-code-1

**Reply by ocots** [reply]
Sinon, il semblerait que l'on puisse générer du html : https://www.juliabloggers.com/local-test-coverage-in-julia/

**Reply by ocots** [reply]
Voir : https://github.com/JuliaCI/LocalCoverage.jl

**Comment by j-l-s** [comment]
julia> using LocalCoverage
julia> Pkg.activate("." )
julia> generate_coverage(; run_test = true)
...
Coverage of /Users/jls/Development/ct/CtBase.jl
```
┌───────────────────────┬───────┬─────┬──────┬──────┐
│ Filename              │ Lines │ Hit │ Miss │    % │
├───────────────────────┼───────┼─────┼──────┼──────┤
│ src/CTBase.jl         │     1 │   1 │    0 │ 100% │
│ src/algorithms.jl     │     0 │   0 │    0 │    - │
│ src/callbacks.jl      │    32 │  32 │    0 │ 100% │
│ src/ctparser.jl       │   358 │ 245 │  113 │  68% │
│ src/ctparser_utils.jl │    60 │  60 │    0 │ 100% │
│ src/default.jl        │    21 │   7 │   14 │  33% │
│ src/descriptions.jl   │    16 │  16 │    0 │ 100% │
│ src/exceptions.jl     │    12 │  10 │    2 │  83% │
│ src/functions.jl      │    96 │  76 │   20 │  79% │
│ src/macros.jl         │    34 │  26 │    8 │  76% │
│ src/model.jl          │   145 │  99 │   46 │  68% │
│ src/plot.jl           │   127 │   0 │  127 │   0% │
│ src/print.jl          │    51 │   0 │   51 │   0% │
│ src/solution.jl       │     2 │   0 │    2 │   0% │
│ src/utils.jl          │    50 │  19 │   31 │  38% │
├───────────────────────┼───────┼─────┼──────┼──────┤
│ TOTAL                 │  1005 │ 591 │  414 │  59% │
└───────────────────────┴───────┴─────┴──────┴──────┘
```

not bad for stats...

**Comment by j-l-s** [comment]
julia> html_coverage(; open = true, dir = tempdir())

will open a new tab/page in browser... (need genhtml to be installed apparently. Use `homebrew` to install it on macOSX
![capture](https://user-images.githubusercontent.com/7069684/227549240-3cf2372c-db93-4a96-8236-7677fb6510ab.jpg)

**Reply by ocots** [reply]
Qu'est-ce que c'est bien `Julia` :-)

**Reply by jbcaillau** [reply]
Very nice indeed. Is it more precise than `codecov` coming with github CI?

**Reply by ocots** [reply]
It's quite the same.

**Reply by ocots** [reply]
I didn't need `genhtml` but `lcov` which was already installed.

**Comment by ocots** [comment]
I have installed the plugin `CodeGPT` for `VScode`, it's amazing. Need to tell.

**Reply by jbcaillau** [reply]
snapshot, please! afraid cannot install it on my old (vs)code server 🤞🏾

**Reply by ocots** [reply]
Ok but this is a very simple example. When I have started to write 

```julia
function Problems()
    for description ∈ examples

end
```

it has proposed what is in gray

![Capture d’écran 2023-03-25 à 22 20 44](https://user-images.githubusercontent.com/66357348/227742585-4143e966-f821-4228-a7b3-d296dad58870.png)

**Reply by jbcaillau** [reply]
Thought you were talking about `codecov` extension for `vscode`.

https://marketplace.visualstudio.com/items?itemName=DanielSanMedium.dscodegpt

**Reply by ocots** [reply]
Actually, I was talking about `GitHub Copilot`: https://marketplace.visualstudio.com/items?itemName=GitHub.copilot
